draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+---
+RFC: XXXX
+Author: Roni Choudhury
+Status: draft
+Created: 2020-08-04
+Last-Modified: 2020-08-04
+---
+
+# RFC XXXX: Multipart Table Upload
+
+This RFC describes how large table data can be safely and efficiently uploaded
+to Multinet, using a variation of the common multipart upload strategy common to
+many web server technologies.
+
+## Motivation
+
+Multipart upload is needed for the following reasons:
+
+1. As a practical matter, our hosting platform Heroku does not allow for
+   requests that take longer than thirty seconds to complete. Some moderately
+   sized CSV files already break this threshold, and so called "big data" tables
+   most certainly will. In order to upload such data files, we will need a way
+   to upload them in chunks.
+2. A long-running one-shot upload does not give the end user much feedback on
+   what is going on. Multipart upload would allow us to provide a more detailed
+   progress meter that can improve the user experience.
+3. If a long-running upload is interrupted for any reason, a multipart upload
+   strategy allows for resuming such uploads without losing the time already
+   spent in uploading one more initial chunks of the file.
+
+## Proposals
+
+### Summary of Approach
+
+The basic idea is as follows: the client initiates a new upload session for a
+given file; this action causes the server to create a new collection in the
+database. Then the client issues one or more "chunk" uploads, each sending a
+range of bytes, up to some size limit, to the server. The server saves each
+chunk as a document in the database collection associated with this file upload.
+Finally, the client issues a "finalize" call, which marks the database
+collection as being complete.
+
+The server can then ask for a file-like object associated with the upload. The
+server can then use this object to perform validation, retrieve table rows, and
+settle them in their final home in the server (as is done currently, directly
+with the file contents).
+
+Once the work is completed, the server can delete the upload collection from the
+database.
+
+### API design
+
+This RFC adds an `uploads` section to the Multinet API, which contains the
+following endpoints:
+
+- `POST /api/uploads` - creates a new upload object by selecting a random,
+non-preexisting SHA string, then creating an Arango collection keyed by the
+string. This endpoint returns the ID string.
+- `POST /api/uploads/<upload-id>/chunk?sequence=<sequence-number>` - uploads one
```

Comment:
It's uncommon to use a querystring for a POST. It would be more standard to have `POST /api/uploads/<upload-id>/chunk/sequence/<sequence-number>`

Reply:
Do you have a reference for this? As a counterexample, the Girder API has plenty of `POST` endpoints that accept query arguments.

The alternative you gave has different RESTful semantics on the face of it: it's saying "create a `sequence` object with ID `<sequence-number>`.

Reply:
This was just my experience. It does seem that it's allowed to do this, just uncommon. See [here](https://stackoverflow.com/questions/611906/http-post-with-url-query-parameters-good-idea-or-not)

Reply:
> This was just my experience. It does seem that it's allowed to do a POST with a query string, just uncommon. See [here](https://stackoverflow.com/questions/611906/http-post-with-url-query-parameters-good-idea-or-not).
> 

One thing that might have gotten lost is that we *are* going to send the chunk data in the body of the request. The SO answer seems concerned with the oddness of not doing that (though, I'm not really sure why that should be so odd). Query arguments seem to be the natural way for us to express the extra data in this case. And I need to update the RFC description with the bit about the request body.

> Your point about about the semantics is correct. Is that not what this endpoint is doing? Are all the sequence numbers determined ahead of time?

Usually in a REST API, the "path elements" in an endpoint refer to actual resources. In this case, we've created an "upload" resource, and with the `/chunk` endpoint we're modifying that object by adding each data chunk. The chunks need to have sequence numbers, but those numbers are like chunk metadata, not resources in and of themselves. That's why I think sending that in a query argument (rather than path elements) is perfectly natural.

Reply:
>One thing that might have gotten lost is that we are going to send the chunk data in the body of the request. 
I understood that. I was just thinking in terms of multi-part forms where all data gets sent in the POST body, not split into the query string and the body

You've convinced me that this makes sense and that the syntax of the requests is appropriate

Reply:
In my experience, the use of query parameters is independent of the HTTP Method used.

Reply:
> Are all the sequence numbers determined ahead of time?

Forgot to answer this above.

I believe the client determines these by matching up slices of the file it's sending with their ordering, so typically you just go `0` for the first chunk, `1` for the second, etc.